motionplan/collision.go

```diff
@@ -248,6 +256,31 @@ func (cg *collisionGraph) collisions() []Collision {
 	return collisions
 }
 
+func (cg *collisionGraph) addCollisionSpecification(specification *Collision) (err error) {
+	if specification.name1 == specification.name2 {
+		return errors.Errorf("cannot add collision specification between entities with the same name: %s", specification.name1)
+	}
+	i, j, ok := cg.getIndices(specification.name1, specification.name2)
+	if !ok {
+		i, j, ok = cg.getIndices(specification.name2, specification.name1)
```

Comment:
Note for @biotinker and @raybjork: On the contradictory constraints concern, if the collision graph is not triangular, can we have a situation where the indices returned here have different penetration depths because the user specified keys and tests in both directions?

Reply:
Penetration depth should be directionality-agnostic; it's "the minimum translation such that objects A and B are not in collision", it's not "how far into B does A penetrate"

Reply:
I was thinking more of a "1" specified for a mapping in one ordering and a "-1" for a mapping in the other ordering.

Reply:
Per my other comment chain here; the `blacklist` functionality should be removed from this PR.

The API-exposed intention of the blacklist isn't "make sure that collisions between these two obstacles are considered", that's the default behavior of an obstacle. It's "when considering collisions between this obstacle and others, consider *only* these others".

So the caller can take that plus a list of all obstacles present, and complement that so `collision.go` only ever consumes "list of collision pairs to ignore". The upstream thing doing the complementing should be the thing that checks for contradictions.

Reply:
Yep.